### PR TITLE
r2r 0.25.1 (new formula)

### PR DESCRIPTION
**Pull request by NolanTrem**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -26,7 +26,7 @@ Thanks for contributing!
 
 * read [the Formula Cookbook](https://docs.brew.sh/Formula-Cookbook) or: `brew create $URL` and make edits
 * `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source foo`
-* `brew audit --new-formula foo`
```
Comment (chenrui333) [inline CONTRIBUTING.md:29]: can you move this change to a separate PR?

File: Formula/r/r2r.rb [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      url "https://files.pythonhosted.org/packages/3d/f9/88d94a75de065ea32619465d2f77b29a0469500e99012523b91cc4141cd1/openpyxl-3.1.5.tar.gz"
+      sha256 "cf0e3cf56142039133628b5acffe8ef0c12bc902d2aadd3e0fe5878dc08d1050"
+    end
+  
+    resource "posthog" do
+      url "https://files.pythonhosted.org/packages/d8/c8/8a7308d5355fedfc400098a75fd191cf615b55aa22ef2a937995326e6f5e/posthog-3.5.0.tar.gz"
+      sha256 "8f7e3b2c6e8714d0c0c542a2109b83a7549f63b7113a133ab2763a89245ef2ef"
+    end
+  
+    resource "pydantic" do
```
Comment (chenrui333) [inline Formula/r/r2r.rb:87]: needs rust dep

File: Formula/r/r2r.rb [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    end
+  
+    def install
+      ENV["PYTHON"] = Formula["python@3.9"].opt_bin/"python3"
+      virtualenv_install_with_resources
+    end
+  
+    test do
+      assert_match "R2R CLI for all core operations", shell_output("#{bin}/r2r --help")
+      system "#{bin}/r2r", "version"
```
Comment (chenrui333) [inline Formula/r/r2r.rb:164]: We need a test that exercises the some of the functionality of the app. Version checks or usage checks (foo --version or foo --help) are not sufficient, as explained in the formula cookbook.

In most cases, a good test would involve running a simple test case: run #{bin}/foo input.txt.

- Then you can check that the output is as expected (with assert_equal or assert_match on the output of shell_output)
- You can also check that an output file was created, if that is expected: assert_predicate testpath/"output.txt", :exist?

Some advice for specific cases:

- If the formula is a library, compile and run some simple code that links against it. It could be taken from upstream's documentation / source examples.
- If the formula is for a GUI program, try to find some function that runs as command-line only, like a format conversion, reading or displaying a config file, etc.
- If the software cannot function without credentials, a test could be to try to connect with invalid credentials (or without credentials) and confirm that it fails as expected.
- Same if the software requires a virtual machine, docker instance, etc. to be running.

**Comment by NolanTrem** [comment]
Thanks for the quick review @chenrui333 didn't realize that the PR into my fork also was included here. This is still a WIP on my end, but will integrate your comments in.

**Comment by chenrui333** [comment]
sounds good, convert it to a draft PR. :)

File: Formula/r/r2r.rb [hunk 4]
```diff
@@ -0,0 +1,167 @@
+class R2r < Formula
+    include Language::Python::Virtualenv
+  
+    desc "Production-ready RAG engine with a sh*t ton of features"
```
Comment (MikeMcQuaid) [inline Formula/r/r2r.rb:4]:
```suggestion
    desc "Production-ready RAG engine with a shit ton of features"
```
or 
```suggestion
    desc "Production-ready RAG engine lots of features"
```

everyone knows what `shi*t` means

Comment (NolanTrem) [reply Formula/r/r2r.rb]: True, we kind of chose this description with that in mind though. Will probably keep it to match our main repo.

File: CONTRIBUTING.md (hunk 1 above)
Comment (NolanTrem) [reply CONTRIBUTING.md]: Moved to #177121

File: Formula/r/r2r.rb [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    homepage "https://r2r-docs.sciphi.ai/"
+    url "https://github.com/SciPhi-AI/R2R/archive/refs/tags/v0.2.51.tar.gz"
+    sha256 "c49299ec90854c7332f2304d7c66c0acae034cf0f20bb01feeb784f6b7b38797"
+    license "MIT"
+  
+    depends_on "python@3.9"
```
[the remark this answers is not in the document]
Comment (NolanTrem) [reply Formula/r/r2r.rb]: Updated to use 3.12

**Review by chenrui333 — COMMENTED** [review]
looks like linux build failed?
